### PR TITLE
Add CT-ness

### PR DESCRIPTION
**Pull request by jcape**
This forward-ports the traits and implementations of the [`mc-crypto-ct-aead`](https://github.com/mobilecoinfoundation/mobilecoin/tree/master/crypto/ct-aead) crate into a fork of RustCrypto's [`aes-gcm`](https://crates.io/crates/aes-gcm), and integrates some misuse-resistence from mobilecoinfoundation/mobilecoin#808

The intent is to publish this to crates.io, and then remove the duplicates from the mobilecoin repo.

### CODE REVIEW

**Review by eranrund — APPROVED** [review]

File: src/lib.rs [hunk 1]
```diff
@@ -96,15 +96,23 @@
 //! [3]: https://research.nccgroup.com/2020/02/26/public-report-rustcrypto-aes-gcm-and-chacha20poly1305-implementation-review/
 //! [4]: https://www.mobilecoin.com/
 
-#![no_std]
+#![cfg_attr(not(feature = "std"), no_std)]
```
Comment (garbageslam) [inline src/lib.rs:99]: I think this is wrong, the crate is unconditionally no_std so why make it more complicated?

File: src/ct.rs [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+        // Conditionally copy the actual plaintext _only_ if the tag verified
+        // correctly, in order to increase misuse resistance and reduce attack
+        // surface for chosen ciphertext attacks.
+        for i in 0..len {
+            buffer[i] = u8::conditional_select(&buffer[i], &ciphertext[i], result);
+        }
+        // Unconditionally zeroize the decryption result to refrain from keeping
+        // a CCA oracle in memory.
+        ciphertext.zeroize();
```
Comment (garbageslam) [inline src/ct.rs:116]: An alternative here is, unconditionally apply keystream again (which is what tony does upstream to conceal the cca) then you don't need a dep on zeroize or an additional optional feature

**Review by garbageslam — CHANGES_REQUESTED** [review]
Looks good other than
(1) let's try to minimize the number of features needed to configure it, and use the same pattern as upstream
(2) I don't think we actually need zeroize here

File: src/lib.rs (hunk 1 above)
Comment (garbageslam) [reply src/lib.rs:99]: We should really try to keep the same feature set as upstream to ease maintenance

Comment (jcape) [reply src/lib.rs:99]: Upstream has a `std` feature, but no internal usage that actually requires `Vec`, that's why it gets away with being unconditionally `no_std`---this isn't adding any additional features to upstream.

File: src/ct.rs (hunk 2 above)
Comment (jcape) [reply src/ct.rs:116]: I didn't add the zeroize feature, that was already there, and when it's enabled, there's some additional benefit to it.

https://github.com/RustCrypto/AEADs/blob/65a7e70c927443213214cb76be43338a167857c8/aes-gcm/src/lib.rs#L215

File: src/lib.rs (hunk 1 above)
Comment (garbageslam) [reply src/lib.rs:99]: i see, i didn't realize upstream has an std feature

File: src/ct.rs (hunk 2 above)
Comment (garbageslam) [reply src/ct.rs:116]: okay i'm sold

**Review by garbageslam — APPROVED** [review]

**Review by christian-oudard — APPROVED** [review]

**Review by isis-mc — COMMENTED** [review]
LGTM

**Pull request merged**